client/src/components/Game/Timer.js

```diff
@@ -0,0 +1,133 @@
+import React, { Component } from 'react';
+import { withFirebase } from '../Firebase';
+import { withRouter } from 'react-router';
+import { compose } from 'recompose';
+import { sendEmail } from '../Helpers'
+
+class Timer extends Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+    }
+  }
+
+  async componentDidMount() {
+    const { id } = this.props.match.params;
+    const game = this.props.firebase.game(id);
+
+    game.get()
+    .then(docSnapshot => {
+      if(docSnapshot.exists) {
+        this.unsubscribe = game.onSnapshot((snapshot) => {
+
+          this.setState({
+            currentPlayerIndex: snapshot.data().currentPlayerIndex,
+            players: snapshot.data().players,
+            gameId: id
+          })
+
+          if(snapshot.data().gameStartTime !== null) {
+            if(snapshot.data().currentPlayerIndex === 0 && this.state.noPlayersSkipped) {
+              this.setState({
+                startTime: snapshot.data().gameStartTime.seconds,
+                timePerTurnInSeconds: parseInt(snapshot.data().timeLimit, 10) * 60,
```

Comment:
I am confused about why this "if" is necessary

Reply:
line 31's if is necessary because for any player after the host, we need to calculate time based on when they opened the app. Originally I was going to use the gameStartTime in conjunction with the backend, but that feature has been pushed back. Also, the noPlayerSkipped is to prevent a situation where the first player is skipped and the second player starts their turn with times up message because their time was calculated based off the game start time and not the time they opened the app.

Reply:
why not just use the else statement and get rid of lines 32-34

Reply:
because lines 32-34 will be used with the backend code and since it didn't change functionality, I didn't see a reason to remove the lines

Reply:
[dead code should be removed](https://stackoverflow.com/questions/15699995/could-someone-explain-the-pros-of-deleting-or-keeping-unused-code)

Reply:
I thought it was considered [dead code](https://en.wikipedia.org/wiki/Dead_code) if it wasn't executed or used in runtime, however this is being used. I understand it currently does the same thing as the code below it but that's only because the backend isn't fully functional yet due to time constraints. If the backend will be added thus requiring a need for the if statement and this code is being used in runtime, what's the reason to remove it?

Reply:
I am stumped as to why the backend would need special behavior specifically if it's the host who has not been skipped..? the backend timer call should start when the host starts the game (?)

Reply:
the backend timer is going to send a field to the front end for when the time the player starts their turn so that the database is the constant source of truth.

Reply:
wait so this isnt special for the first player && notSkipped

Reply:
it is because instead of the second startTime getting a date it will grab the turnStartTime. we don't want the host grabbing the turnTime because that would be null when the game first starts. my idea is that when an email is sent (meaning a turn has ended) that turnStartTime would be updated by the backend for the front end to grab

Reply:
ohh i see ur line of thinking now, thanks for explaining
you can just update that 1st time when the host [clicks the start game](https://github.com/googleinterns/Pictophone/pull/68#discussion_r465300068) button, more straightforward and dont need to implement extra checks

Reply:
that's what i plan on doing, but the backend is going to have to keep updating that field. it can't be updated in the front end unless we send a response back to the front end saying "email sent", however even then i'd still need to implement some sort of check because it takes a second for the timestamp field to populate. So if i update it from the backend, then when the page renders, it checks for if it's the host or another player, it'll know which state to set

Reply:
from a design perspective i would expect sendEmail() to do exactly that -- send the email. I wouldn't expect it to do side effects like changing the timer. Better to break these two functionalities into two different functions (sendEmail() and, say, newTime().. this naming is horrible but u get the idea) that can call each other and also be individually called from the frontend. A separation of concerns and better modularity removes the problem of one function having undesired side effects.

WITH THAT being said, I don't think it's worth to fix given the timeline and you should probably go ahead and merge it anyways; we can improve the code design at a later date

Reply:
i appreciate the at length discussion, you're right. Maybe a call to a method in the timer class?? once i fix the timer resetting i will merge

Reply:
..I haven't thought about this design at length so it def has drawbacks, but intuitively I would implement a startGame api call and a nextTurn api call. some sort of setTime() method in the timer class, yeah, that both Rest controllers would call. and then nextTurn would also call the sendEmail() method, though startGame would not. This abstracts away our implementation detail to have more simple api calls. Also offers flexibility for the future, should we want to move game-advancing logic to the backend to offer more user security (so they cannot mess with our Firestore game array, which is currently open to read/write from any auth user). In that case, we can just write more game logic in the nextTurn call.